builds/jenkins-build-sls-app/Jenkinsfile

```diff
@@ -0,0 +1,1062 @@
+        } catch(ex) {
+          def outputLog = readFile('output.log').trim()
+          echo "template validation log $outputLog"
+          throw ex
+        }
+        def outputLog = readFile('output.log').trim()
+        echo "template validation log $outputLog"
+
+        def cft = readFile('./.serverless/cloudformation-template-update-stack.json').trim()
+        def cftJson = parseJson(cft)
+        def outstandingResources = whiteListModule.validate(cftJson)
+        if(!outstandingResources.isEmpty()) throw new Exception("Invalid CloudFormation template. The resource types not allowed: ${outstandingResources}")
+      }
+    }
+
+    stage("Deployment to ${env_key} environment") {
+      events.sendStartedEvent('DEPLOY_TO_AWS', 'Deployment started to staging AWS environment', context_map, environment_logical_id)
+      events.sendStartedEvent('UPDATE_ENVIRONMENT', "Environment status update event for ${env_key} deployment", environmentDeploymentMetadata.generateEnvironmentMap("deployment_started", environment_logical_id, null), environment_logical_id)
+
+      echo "starts Deployment to ${env_key} Env"
+      def lambdaARN = null
+      environmentDeploymentMetadata.setEnvironmentEndpoint(lambdaARN)
+
+      events.sendStartedEvent('UPDATE_ENVIRONMENT', "Environment status update event for ${env_key} deployment", environmentDeploymentMetadata.generateEnvironmentMap("deployment_started", environment_logical_id, null), environment_logical_id)
+
+      withCredentials([
+        [$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: configLoader.AWS_CREDENTIAL_ID, secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']
+      ]) {
+        try {
+          // initialize aws credentials
+          def randomString = utilModule.generateRequestId();
+          def credsId = "jazz-${randomString}";
+          sh "aws configure set profile.${credsId}.region ${configLoader.AWS.REGION}"
+          sh "aws configure set profile.${credsId}.aws_access_key_id $AWS_ACCESS_KEY_ID"
+          sh "aws configure set profile.${credsId}.aws_secret_access_key $AWS_SECRET_ACCESS_KEY"
+
+          // Generate serverless yml file with domain added in function name
+          echo "Generate deployment env with domain"
+
+          if (isScheduleEnabled) {
+            def eventsArns = getEventsArn(config, environment_logical_id)
+            for (def i = 0; i < eventsArns.size(); i++) {
+              def arn = eventsArns[i]
+              events.sendCompletedEvent('CREATE_ASSET', null, utilModule.generateAssetMap("aws", arn, "cloudwatch_event", config), environment_logical_id);
+            }
+          }
+
+          def envBucketKey = "${env_key}${configLoader.JAZZ.S3_BUCKET_NAME_SUFFIX}"
+          echoServerlessFile()
+          def deployOutput = servelessDeploy(environment_logical_id, envBucketKey)
+
+// Shoving deploy output lines into array
+          def deployLines = deployOutput.split('\n')
```

Comment:
Handle unsuccessful build scenarios & send appropriate events to mark the deployment as failed.

Reply:
Well, here the unsuccessful build scenarios have already been handled in the existing catch block as follows:
        try {
          // initialize aws credentials
          def randomString = utilModule.generateRequestId();
          def credsId = "jazz-${randomString}";
.................................
          echoServerlessFile()
          def deployOutput = servelessDeploy(environment_logical_id, envBucketKey)
.................................
          send_status_email(config, 'COMPLETED', svc_response)

        } catch (ex) {
          send_status_email(config, 'FAILED', '')
          events.sendFailureEvent('UPDATE_ENVIRONMENT', ex.getMessage(), environmentDeploymentMetadata.generateEnvironmentMap("deployment_failed", environment_logical_id, null), environment_logical_id)
          events.sendFailureEvent('UPDATE_DEPLOYMENT', ex.getMessage(), environmentDeploymentMetadata.generateDeploymentMap("failed", environment_logical_id, gitCommitHash), environment_logical_id)
          events.sendFailureEvent('DEPLOY_TO_AWS', ex.getMessage(), context_map, environment_logical_id)
          error ex.getMessage()
        }